### PR TITLE
refactor(medusa): Cleanup and fix CartService

### PR DESCRIPTION
**Pull request by adrien2p**

### CODE REVIEW

**Review by srindom — CHANGES_REQUESTED** [review]
Nice work @adrien2p! Have a few todos for you :)

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:610]: Perfect, I was wondering a way to be sure to run in the transaction without having to clone the service unnecessarily :) thank you very much for the explaination. This was a point I wanted to discuss to be sure I did understand the logic without breaking everything 😂

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts:891]: sorry @srindom but for me that method is not private since by default a class member is public unless we specify that it is private :) 
I can't not refer to the name as well because it is container `_`. But event `atomicPhase_` contains the `_` but this member is either public or protected otherwise it would not be accessible from the derivates class

Comment (adrien2p) [reply packages/medusa/src/services/cart.ts]: How do I identify a private method agains public/protected?

Comment (srindom) [reply packages/medusa/src/services/cart.ts]: yes - exactly; the suffix indicates that it is private. 

I am aware that it is not technically private, but using the `_` consumers are at least aware that this method is not supposed to be called from outside of the method.

Comment (srindom) [reply packages/medusa/src/services/cart.ts]: The convention is inspired by Google: https://google.github.io/styleguide/jsguide.html#naming-method-names

Comment (srindom) [reply packages/medusa/src/services/cart.ts]: We may want to revise this, however, given the migration to TS

File: packages/medusa/src/services/cart.ts
Comment (adrien2p) [reply packages/medusa/src/services/cart.ts]: @srindom We are just waiting for that point if I am right?

Comment (srindom) [reply packages/medusa/src/services/cart.ts]: @zakariaelas's work is coming in: #1319 let's wait until that is merged? Wdyt?

Comment (adrien2p) [reply packages/medusa/src/services/cart.ts]: @srindom yeah alright, I will manage the conflict later :)

**Review by srindom — APPROVED** [review]
LGTM! 🎉

File: packages/medusa/src/services/cart.ts
Comment (anup-a) [inline packages/medusa/src/services/cart.ts:818]: Hey @adrien2p , it seems like because of this line, we are not able to remove discount from a cart.

**Pull request merged**
